pkg/networkmanager/firewall-client.es6

```diff
@@ -0,0 +1,198 @@
+    let installed = !!firewalld_service.exists;
+
+    if (firewall.installed == installed)
+        return;
+
+    firewall.installed = installed;
+    firewall.dispatchEvent('changed');
+});
+
+function fetchServiceInfos(services) {
+    return cockpit.all(services.map(service => {
+        if (firewall.services[service])
+            return firewall.services[service];
+
+        return firewalld_dbus.call('/org/fedoraproject/FirewallD1',
+                                   'org.fedoraproject.FirewallD1',
+                                   'getServiceSettings', [service])
+                .then(reply => {
+                    const [ , name, description, ports ] = reply[0];
+
+                    let info = {
+                        id: service,
+                        name: name,
+                        description: description,
+                        ports: ports.map(p => ({ port: p[0], protocol: p[1] }))
+                    };
+
+                    firewall.services[service] = info;
+                    return info;
+                });
+    }))
+
+    /*
+     * Work around `cockpit.all()` returning results in individual arguments -
+     * that's just confusing and doesn't work with ES6 style functions.
+     */
+            .then(function () {
+                return Array.prototype.slice.call(arguments);
+            });
+}
+
+firewalld_dbus.addEventListener('owner', (event, owner) => {
+    firewall.enabled = !!owner;
+
+    if (!firewall.enabled) {
+        firewall.services = {};
+        firewall.enabledServices = new Set();
+        firewall.dispatchEvent('changed');
+        return;
+    }
+
+    firewalld_dbus.call('/org/fedoraproject/FirewallD1',
+                        'org.fedoraproject.FirewallD1.zone',
+                        'getServices', [''])
+            .then(reply => fetchServiceInfos(reply[0]))
+            .then(services => services.map(s => firewall.enabledServices.add(s.id)))
+            .then(() => firewall.dispatchEvent('changed'))
+            .catch(error => console.warn(error));
+});
+
+firewalld_dbus.subscribe({
+    interface: 'org.fedoraproject.FirewallD1.zone',
+    path: '/org/fedoraproject/FirewallD1',
+    member: 'ServiceAdded'
+}, (path, iface, signal, args) => {
+    const service = args[1];
+
+    fetchServiceInfos([service])
+            .then(info => {
+                firewall.enabledServices.add(info[0].id);
+                firewall.dispatchEvent('changed');
+            })
+            .catch(error => console.warn(error));
+});
+
+firewalld_dbus.subscribe({
+    interface: 'org.fedoraproject.FirewallD1.zone',
+    path: '/org/fedoraproject/FirewallD1',
+    member: 'ServiceRemoved'
+}, (path, iface, signal, args) => {
+    const service = args[1];
+
+    firewall.enabledServices.delete(service);
+    firewall.dispatchEvent('changed');
+});
+
+cockpit.spawn(['sh', '-c', 'pkcheck --action-id org.fedoraproject.FirewallD1.all --process $$ -u 2>&1'])
```

Comment:
Maybe replace `-u` with `--allow-user-interaction` -- short options are nice for users to type; but not really in code imo.